tsdb/engine/tsm1/engine.go

```diff
@@ -653,6 +653,162 @@ func (e *Engine) Backup(w io.Writer, basePath string, since time.Time) error {
+	if err != nil {
+		return err
+	}
+
+	if err := e.index.SnapshotTo(path); err != nil {
+		return err
+	}
+
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+
+	// Remove the temporary snapshot dir
+	defer os.RemoveAll(path)
+
+	// Recursively read all files from path.
+	files, err := readDir(path, "")
+	if err != nil {
+		return err
+	}
+
+	// filtered is different here, these are files we need to trim based on timestamps.
+	// whole is an optimisation: if a TSM is fully contained in the range, then we keep the whole thing.
+	var filtered []string
+	var whole []string
+	for _, file := range files {
+		if !strings.HasSuffix(file, ".tsm") {
+			continue
+		}
+
+		var tombstonePath string
+		f, err := os.Open(filepath.Join(path, file))
+		if err != nil {
+			return err
+		}
+		r, err := NewTSMReader(f)
+		if err != nil {
+			return err
+		}
+
+		// Grab the tombstone file if one exists.
+		if r.HasTombstones() {
+			tombstonePath = filepath.Base(r.TombstoneFiles()[0].Path)
+		}
+
+		// Skip this file if no overlap with start/end
+		min, max := r.TimeRange()
+		if min > end.UnixNano() || max < start.UnixNano() {
+			r.Close()
+			continue
+		}
+
+		r.Close()
+
+		if tombstonePath != "" {
+			whole = append(filtered, tombstonePath)
+		}
+
+		if min > start.UnixNano() && max < end.UnixNano() {
+			whole = append(filtered, file)
+		} else {
+			// at this point, we know we're not totally outside or inside start and end so we'll need to scan the blocks
+			filtered = append(filtered, file)
+		}
+	}
+
+	for _, f := range filtered {
+		if err := e.filterFileToBackup(f, basePath, filepath.Join(path, f), start.UnixNano(), end.UnixNano(), tw); err != nil {
+			return err
+		}
+	}
+
+	for _, f := range whole {
+		if err := e.writeFileToBackup(f, basePath, filepath.Join(path, f), tw); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *Engine) filterFileToBackup(name, shardRelativePath, fullPath string, start, end int64, tw *tar.Writer) error {
+	f, err := os.Stat(fullPath)
+	if err != nil {
+		return err
+	}
+
+	h := &tar.Header{
+		Name:    filepath.ToSlash(filepath.Join(shardRelativePath, name)),
+		ModTime: f.ModTime(),
+		Size:    f.Size(),
+		Mode:    int64(f.Mode()),
+	}
+	if err := tw.WriteHeader(h); err != nil {
+		return err
+	}
+	fr, err := os.Open(fullPath)
+	if err != nil {
+		return err
+	}
+
+	defer fr.Close()
+
+	// TODO (Adam):  instead of io.copy as in writeFileToBackup, we will need to:
+
+	// 2.  write the blocks we want to keep
+	// 3.  write an index?????
+	// 4.  write a footer
+
+	// will probably need most of this
+	r, err := NewTSMReader(fr)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	w, err := NewTSMWriter(tw)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		closeErr := w.Close()
+		if err == nil {
+			err = closeErr
+		}
+
+		// Check for errors where we should not remove the file
+		// REVIEWER:  do I need to check anything here? lifted this code from compact.go
+
+		if err != nil {
+			w.Remove()
```

Comment:
This isn't necessary as you are writing a TSM file to a tar stream.  This would attempt to delete it from the tar stream which you can't really do.